ethereum/core/src/main/java/tech/pegasys/teku/core/epoch/status/ValidatorStatuses.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.core.epoch.status;
+
+import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.get_block_root_at_slot;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.datastructures.operations.AttestationData;
+import tech.pegasys.teku.datastructures.state.BeaconState;
+import tech.pegasys.teku.datastructures.state.Checkpoint;
+import tech.pegasys.teku.datastructures.state.Validator;
+import tech.pegasys.teku.datastructures.util.AttestationUtil;
+import tech.pegasys.teku.datastructures.util.BeaconStateUtil;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.ssz.SSZTypes.SSZList;
+
+public class ValidatorStatuses {
+  private final List<ValidatorStatus> statuses;
+  private final TotalBalances totalBalances;
+
+  private ValidatorStatuses(
+      final List<ValidatorStatus> statuses, final TotalBalances totalBalances) {
+    this.statuses = statuses;
+    this.totalBalances = totalBalances;
+  }
+
+  public static ValidatorStatuses create(final BeaconState state) {
+    final SSZList<Validator> validators = state.getValidators();
+
+    final UInt64 currentEpoch = BeaconStateUtil.get_current_epoch(state);
+    final UInt64 previousEpoch = BeaconStateUtil.get_previous_epoch(state);
+
+    final List<ValidatorStatus> statuses =
+        validators.stream()
+            .map(validator -> ValidatorStatus.create(validator, previousEpoch, currentEpoch))
+            .collect(Collectors.toCollection(() -> new ArrayList<>(validators.size())));
```

Comment:
this might be just a java oddity, but why not just `Collectors.toList()`?

Reply:
If we just used `toList()` it would create a default sized backing array and then when the number of validators outgrows it have to copy them all to a new, bigger list.  It will wind up repeating that process at least a few times before it is actually big enough to hold the validators, and that final array may actually wind up being much bigger than we needed, wasting memory.

Since we know the final list size we'll need and it's a really big number that we want to avoid having to copy, creating the array this way winds up more efficient in CPU and memory usage.